edocumentation-domain/src/main/java/org/dcsa/edocumentation/domain/persistence/state/BookingStateMachineImpl.java

```diff
@@ -0,0 +1,85 @@
+package org.dcsa.edocumentation.domain.persistence.state;
+
+import lombok.Getter;
+import org.dcsa.edocumentation.domain.persistence.entity.enums.BkgDocumentStatus;
+
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static org.dcsa.edocumentation.domain.persistence.state.BookingStateMachineTables.BOOKING_STATE_TRANSITION_TABLE;
+import static org.dcsa.edocumentation.domain.persistence.state.BookingStateMachineTables.INITIAL_STATE;
+
+public class BookingStateMachineImpl implements BookingStateMachine {
+
+  @Getter
+  BkgDocumentStatus currentStatus;
+
+  BookingStateMachineImpl() {
+    currentStatus = INITIAL_STATE;
+  }
+
+  BookingStateMachineImpl(BkgDocumentStatus currentStatus) {
+    this.currentStatus = Objects.requireNonNull(currentStatus, "currentStatus cannot be null");
+    verifyState(currentStatus);
+  }
+
+  private void transitionTo(BkgDocumentStatus successor) {
+    verifyTransition(successor);
+    currentStatus = successor;
+  }
+
+  void verifyTransition(BkgDocumentStatus intendedSuccessor) {
+    Objects.requireNonNull(intendedSuccessor, "successor status cannot be null");
+    Set<BkgDocumentStatus> successors = BOOKING_STATE_TRANSITION_TABLE.get(this.currentStatus);
+    if (successors == null) {
```

Comment:
mixing Objects.requireNonNull and if( successors == null) feels a bit weird

Reply:
For me, this is not an issue. The `Objects.requireNonNull` is an easy way to ensure that input it is not not-null and given a good error message for that. It is an NPE but I feel that is acceptable here if it avoids an explicit `if`.

The `if (successors == null) {`-check is really about the state machine likely being incomplete, so a NPE feels wrong.